src/__test__/index.test.ts

```diff
@@ -333,21 +368,242 @@ describe('test rpc requests to Snap', () => {
+			expect(ui).toRender(
+				panel([
+					heading(prompt),
+					divider(),
+					...description.map((t) => text(t)),
+				]),
+			);
+
+			await ui.ok();
+
+			const initPairingJson: any = (await initPairingReq).response;
+			const initPairingResult =
+				initPairingJson.result as InitPairingResponse;
+			const qrCode = initPairingResult.qrCode;
+			const qrCodeObj = JSON.parse(qrCode) as QrCode;
+
+			expect(qrCodeObj.pairingId).toEqual(expect.any(String));
+			expect(qrCodeObj.webEncPublicKey).toEqual(expect.any(String));
+			expect(qrCodeObj.signPublicKey).toEqual(expect.any(String));
+
+			// Test run pairing
+			await simulator.signIn();
+			await simulator.pairing(qrCode);
+
+			const runPairingReq = request({
+				method: InternalMethod.TssRunPairing,
+				origin: ORIGIN,
+			});
+
+			const runPairingJson: any = (await runPairingReq).response;
+			const runPairingResult =
+				runPairingJson.result as RunPairingResponse;
+			expect(runPairingResult.deviceName).toEqual(DEVICE_NAME);
+			expect(runPairingResult.address).toBeNull();
+
+			// Test key generation
+			const keygenReq = request({
+				method: InternalMethod.TssRunKeygen,
+				origin: ORIGIN,
+			});
+			await simulator.keygen();
+
+			const keyGenJson: any = (await keygenReq).response;
+			const runKeyGenResult = keyGenJson.result as RunKeygenResponse;
+			expect(runKeyGenResult.address).toEqual(expect.any(String));
+
+			request({
+				method: InternalMethod.TssRunBackup,
 				origin: ORIGIN,
 			});
-			await unpairReq;
+
+			// Test signing
+			const keyshareReq = request({
+				method: InternalMethod.E2eTestGetKeyShare,
+				origin: ORIGIN,
+			});
+			const keyshareReqJson: any = (await keyshareReq).response;
+			const keyshareResult = keyshareReqJson.result as {
+				distributedKey: DistributedKey;
+				pairingData: PairingData;
+			};
+
+			const accountId = keyshareResult.distributedKey.accountId;
+			const keyring = new SimpleKeyring({
+				wallets: {
+					[accountId]: {
+						...MOCK_WALLET_ACCOUNT,
+						distributedKey: keyshareResult.distributedKey,
+					},
+				},
+				requests: {},
+			});
+
+			const pairingData = keyshareResult.pairingData;
+			const runTssSign = genMockRunTssSign(pairingData);
+
+			// Personal sign
+			let personalSignResult: any = null;
+
+			keyring
+				.signPersonalMessage(
+					WALLER_ADDRESS,
+					mockPersonalMsg,
+					runTssSign,
+				)
+				.then((resp: any) => {
+					personalSignResult = resp;
+				})
+				.catch((err) => {
+					console.log('err', err);
+				});
+			simulator.sign();
+			while (!personalSignResult) {
+				await delay(0);
+				console.log('waiting for personalSignResult');
+			}
+
+			await simulator.cleanUpSimulation();
 		});
 
-		it('repairing: without tss_runRePairing should be success', async () => {
+		it('tss_initPairing, tss_runPairing, tss_runKeygen, V4 typed signing should be success', async () => {
 			const { request } = await installSnap();
+			// Test init pairing
```

Comment:
We are do pairing for every test case, can we do it once or if we want to do new pairing each time then can we extract the duplicate code and make a function so we can call it everywhere.

Reply:
I faced some issue with low-level library, let me try to find a solution for this.